### PR TITLE
Incubating WebSocketNetworkTransport

### PR DESCRIPTION
**Pull request by martinbonnin**
Introduce a new `NetworkTransport` for WebSockets that rewrites most of the internal machinery:
- Uses a callbackFlow instead of a multiple Channels of messages
- Simplify the WsProtocol interface
- Remove ktor as a dependency

This fixes a couple of issues:
- https://github.com/apollographql/apollo-kotlin/issues/5217
- https://github.com/apollographql/apollo-kotlin/issues/4652

And should make debugging easier for other ones like:
- https://github.com/apollographql/apollo-kotlin/issues/3020
- https://github.com/apollographql/apollo-kotlin/issues/2765

The new `NetworkTransport` is named `WebSocketNetworkTransport`, just like the current one. The only thing that changes is the package name, from `com.apollographql.apollo3.network.ws` to `com.apollographql.apollo3.network.websocket`. The current plan is to:

1. Introduce this as a separate artifact for the next version (`apollo-websocket-network-transport-incubating`). 
2. Make it the default for v4. The breaking change will be:
  * the package name
  * the `WsProtocol` API that is wildly changed. Anyone using a custom `WsProtocol` will need to rewrite some code but I think this is for the greater good.

Opening as draft the time to re-read everything and add KDoc

Relates to #5648

### CODE REVIEW

File: libraries/apollo-websocket-network-transport-incubating/src/jsMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketEngine.js.kt [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+                    && window.process.versions != null 
+                    && window.process.versions.node != null)
+                """
+  ) as Boolean
+}
+
+/*
+ * The function below works due to ktor being used for the HTTP engine
+ * and how ktor imports ws, if this changes, ws would need to be added
+ * as a direct dependency.
```
Comment (BoD) [inline libraries/apollo-websocket-network-transport-incubating/src/jsMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketEngine.js.kt:142]: I think this comment can just be removed?

File: libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/SubscribableWebSocket.kt [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+
+  /**
+   * A network error happened and this operation should be restarted
+   */
+  fun onRetry(throwable: Throwable)
+}
+
+/**
+ * @param terminated a state that signals that a general error happened all future WebSocket messages
+ * must be ignored. This is to make it robust to [WsProtocol] that send spurious messages.
```
Comment (BoD) [inline libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/SubscribableWebSocket.kt:60]:
```suggestion
 * @param terminated a state that signals that a general error happened and all future WebSocket messages
 * must be ignored. This is to make it robust to [WsProtocol] that send spurious messages.
```

File: libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketNetworkTransport.kt [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    private val headers: List<HttpHeader>,
+    private val webSocketEngine: WebSocketEngine,
+    private val idleTimeoutMillis: Long,
+    private val wsProtocolBuilder: WsProtocol.Builder,
+    private val reopenWhen: (suspend (Throwable, attempt: Long) -> Boolean),
+    private val pingIntervalMillis: Long,
+    private val connectionAcknowledgeTimeoutMillis: Long,
+) : NetworkTransport {
+
+  private val isConnectedPrivate = MutableStateFlow(false)
```
[the remark this answers is not in the document]
Comment (martinbonnin) [reply libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketNetworkTransport.kt:40]: Do you have any reference?

File: libraries/apollo-websocket-network-transport-incubating/src/jsMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketEngine.js.kt (hunk 1 above)
Comment (martinbonnin) [reply libraries/apollo-websocket-network-transport-incubating/src/jsMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketEngine.js.kt:142]: 100%

File: libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/AppSyncWsProtocol.kt [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+import com.apollographql.apollo3.api.json.readAny
+import com.apollographql.apollo3.api.json.writeAny
+import com.apollographql.apollo3.api.toJsonString
+import okio.Buffer
+
+class AppSyncWsProtocol(
+    val authorization: suspend () -> Any? = { null },
+) : WsProtocol {
+  override val name: String
+    get() = "graphql-ws"
```
[the remark this answers is not in the document]
Comment (BoD) [reply libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/AppSyncWsProtocol.kt:19]: wow! TIL :)

File: libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketNetworkTransport.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (BoD) [reply libraries/apollo-websocket-network-transport-incubating/src/commonMain/kotlin/com/apollographql/apollo3/network/websocket/WebSocketNetworkTransport.kt:40]: Yes it's hinted at [here](https://kotlinlang.org/docs/coding-conventions.html#names-for-backing-properties)

**Comment by martinbonnin** [comment]
Obsolete

**Pull request closed without merging**